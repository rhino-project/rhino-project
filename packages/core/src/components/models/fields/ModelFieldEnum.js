import PropTypes from 'prop-types';
import { useCallback, useMemo } from 'react';
import FieldSelectControlled from 'rhino/components/forms/fields/FieldSelectControlled';
import { useGlobalComponentForAttribute } from 'rhino/hooks/overrides';
import { useModelAndAttributeFromPath } from 'rhino/hooks/models';
import { enumFromIndexWithTitle } from 'rhino/utils/ui';

export const ModelFieldEnumBase = ({ model, ...props }) => {
  const { path } = props;
  const { attribute } = useModelAndAttributeFromPath(model, path);

  const options = useMemo(
    () =>
      enumFromIndexWithTitle(attribute?.enum, `${attribute.readableName}...`),
    [attribute]
  );

  const accessor = useCallback((value) => value || -1, []);

  return (
    <FieldSelectControlled accessor={accessor} {...props}>
      {options}
    </FieldSelectControlled>
  );
};

ModelFieldEnumBase.propTypes = {
  model: PropTypes.oneOfType([PropTypes.string, PropTypes.object]).isRequired,
  path: PropTypes.string.isRequired
};

const ModelFieldEnum = (props) =>
  useGlobalComponentForAttribute('ModelFieldEnum', ModelFieldEnumBase, props);

export default ModelFieldEnum;
