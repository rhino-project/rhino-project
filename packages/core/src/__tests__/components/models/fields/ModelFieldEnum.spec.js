import { sharedFieldTests } from './sharedFieldTests';
import ModelFieldEnum from 'rhino/components/models/fields/ModelFieldEnum';

describe('ModelFieldEnum', () => {
  sharedFieldTests(ModelFieldEnum);
});
