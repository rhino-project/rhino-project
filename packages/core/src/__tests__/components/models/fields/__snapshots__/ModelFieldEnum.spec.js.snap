// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModelFieldEnum should render with global override shorthand 1`] = `
<DocumentFragment>
  <div>
    Bar
  </div>
</DocumentFragment>
`;

exports[`ModelFieldEnum should render with global override shorthand for model 1`] = `
<DocumentFragment>
  <div>
    Bar
  </div>
</DocumentFragment>
`;

exports[`ModelFieldEnum should render with global override shorthand for model and attribute 1`] = `
<DocumentFragment>
  <div>
    Bar
  </div>
</DocumentFragment>
`;
